Added option to lookup_name to raise error

ayush1999 opened the pull request:
Solves: #6740 

Added a keyword argument named `raises` ( default value is False ).
When raises is False, the function works just like before.
When raises is True, the function raises an exception if the name(alias) isn't found.

mirca commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,34 @@ def get_transform(self, fromsys, tosys):
         """
-
-        return self._cached_names.get(name, None)
+        if (raises == False):
```
Not need to put the expression under parenthesis. Can you also use ``if not raises:`` for PEP8 reasons*?

*: https://docs.quantifiedcode.com/python-anti-patterns/readability/comparison_to_true.html

mirca commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,34 @@ def get_transform(self, fromsys, tosys):
+        if (raises == False):
+            return self._cached_names.get(name, None)
+        else:
+            if self._cached_names.get(name, None) == None:
```
Same thing here: ``self._cached_names.get(name, None) == None`` ---> ``self._cached_names.get(name, None) is None``

mirca commented:
BTW, this reference might help you write a proper unit test for this feature request: https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions

astrofrog commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,34 @@ def get_transform(self, fromsys, tosys):
+            if self._cached_names.get(name, None) is None:
+                raise Exception("name not found")
+            else:
+                return self._cached_names.get(name, None)
```
It might be cleaner to simply do:

```
if name in self._cached_names:
    return return self._cached_names[name]
elif raises:
    raise Exception(...)
```

I think we might also want a better keyword name than ``raises``, but I'll let the astropy.coordinates maintainers comment on this.

ayush1999 replied:
@astrofrog Sure, I'll make the above mentioned changes. Also, can you please tell me which test file should be amended?

astrofrog replied:
The test would need to be added here: https://github.com/astropy/astropy/blob/master/astropy/coordinates/tests/test_transformations.py

In general if you are looking at file ``astropy/package/blah.py``, there should usually be a matching ``astropy/package/tests/test_blah.py``.

ayush1999 replied:
@astrofrog  Thanks! I'll add the tests. Also, can you please tell me why is the travis test failing? The reason seems unclear.

MSeifert04 replied:
The failure is unrelated and just a web-resource that is temporarily unavailable. When you push the test(s) it will rebuild and (probably) pass.

ayush1999 commented:
Can someone please guide me in writing a proper test for this? Thanks!

bsipocz commented:
Hi @ayush1999!

you need to add a tests that tests for the example of #6740, testing both when the new keyword arg is True and once when False. The first should get an exception, the latter the previous behaviour.

But. As it was said before, please follow PEP8 in your commits, they are even picked up by CI. (easiest to check is with running ``pycodestyle`` on the files you modified). 
I was also advised above to use a different name for the new argument than ``raises``. Generally it is a very bad idea to overwrite standard python statements with variables and arguments without a very strong reason.

mirca commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,34 @@ def get_transform(self, fromsys, tosys):
+            return self._cached_names.get(name, None)
+        else:
+           if name in self._cached_names:
+                      return self._cached_names[name]
```
Indentations here and below do not look correct.

ayush1999 commented:
@mirca  @astrofrog Please go through these changes and tell me if any more changes are required. I've changed the name of the argument from `raises` to `raise_exception` and added a test case identical to the one raised in the issue #6740 . Also, Can you please tell me why does the `astropy-bot` test fail? Thanks! :)

MSeifert04 commented on astropy/coordinates/tests/test_transformations.py:
```diff
@@ -430,3 +430,9 @@ def test_function_transform_with_differentials():
+def test_lookup_name():
+    g = t.TransformGraph()
+    with pytest.raises(NameError):
+            g.lookup_name('banana', raise_exception= True)
```
Could you use 4 spaces as indentation here? And for keyword-arguments there's generally no whitespace around the`=`.

MSeifert04 commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,34 @@ def get_transform(self, fromsys, tosys):
         return self._composite_cache[fttuple]
 
-    def lookup_name(self, name):
+    def lookup_name(self, name, raise_exception=False):
```
I would prefer an argument name that is more descriptive, maybe something like `fail_if_none`?

MSeifert04 commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,34 @@ def get_transform(self, fromsys, tosys):
         name : str
             The alias to look up.
-
+        raise_exception : Bool (default value is False)
```
It's lowercase `bool`.

MSeifert04 commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,34 @@ def get_transform(self, fromsys, tosys):
+            if name in self._cached_names:
+                return self._cached_names[name]
+            else:
+                raise NameError("name not found")
```
Not sure if a `NameError` is appropriate here, `NameError`s are for accesses to undefined variables. Maybe just use `return self._cached_names` and just let the `KeyError` "happen".

MSeifert04 commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,34 @@ def get_transform(self, fromsys, tosys):
         Returns
         -------
         coordcls
+        if raise_exception == False:
```
How does this look in the sphinx build?

bsipocz replied:
I was very uncomfortable with this, too, we don't do this format anywhere else. 

There is a sphinx section for exceptions called ``Raises``, I think that would be a good place to add this info.

```
Raises
------

KeyError
    If the coordinate class corresponding to the ``name`` does not exists
    and ``nameofnewargument`` is True.
```

MSeifert04 commented:
And it needs a changelog entry :)

MSeifert04 commented:
Could you also put a changelog entry in `CHANGES.rst` in the project root directory?

@ayush1999 Do you know how to squash commits? And thank you very much for your contribution!

MSeifert04 commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,41 @@ def get_transform(self, fromsys, tosys):
         return self._composite_cache[fttuple]
 
-    def lookup_name(self, name):
+    def lookup_name(self, name, suppress_error=False):
```
The default should be `True` to keep it backwards-compatible.

ayush1999 commented:
Sorry for the confusion, @MSeifert04 . I hope I've made all necessary changes. I'll change the `CHANGES.rst` file once all changes have been approved. Thanks!

MSeifert04 commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,41 @@ def get_transform(self, fromsys, tosys):
+            if name in self._cached_names:
+                return self._cached_names[name]
+            else:
+                raise KeyError("name not found")
```
You don't need to raise it explicitly, you can simply use

```
...
else:
    return self._cached_names[name]
```
That will raise the KeyError automatically if the `name` is not inside the `_cached_names`.

MSeifert04 commented:
No need to wait, the changelog entry has to be reviewed too :)

However https://github.com/astropy/astropy/pull/6747#discussion_r146134126 should be adressed. I don't think the rendered documentation will look expected - could you check that? And do you know how to squash commits?

ayush1999 commented:
@MSeifert04 Can you please tell me how can I address the comment? What about squashing commits?

MSeifert04 commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,38 @@ def get_transform(self, fromsys, tosys):
             The alias to look up.
+        suppress_error : bool (default value is False)
+            True : doesn't raise exception if 'name' isn't found. (returns None)
+            False : raises exception if 'name' isn't found.
```
Could you also add a `.. versionadded:: 3.0` to the parameter description?

And I think it would be better as a sentence like `If True (the default) then it returns None if the name isn't found otherwise it will raise an Exception`. Sphinx generally doesn't respect newlines so this could look weird. You may need to build the documentation locally to see if it looks alright.

MSeifert04 commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,38 @@ def get_transform(self, fromsys, tosys):
         ----------
         name : str
             The alias to look up.
+        suppress_error : bool (default value is False)
```
I don't know if we should provide the `(default value is False)` here. I think that belongs into the description.

MSeifert04 commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,38 @@ def get_transform(self, fromsys, tosys):
         Returns
         -------
         coordcls
+        if suppress_error == True:
```
> Can you please tell me how can I address the comment?

The easy way:

Just make it text: "The coordinate class corresponding to the ``name``. If no such class exists it returns None if suppress_error is true, otherwise it raises an exception."

The complicated way:

Build the docs locally and check how they look. If they look alright you can post it here as image, if it doesn't look okay, then change it _until_ it looks good. :)

MSeifert04 commented:
> What about squashing commits?

Just wanted to know if you know how to do that?

ayush1999 commented:
@MSeifert04 Sorry for the confusion. Next PR onwards, I'll definitely keep on amending the same commit. :)

MSeifert04 commented on CHANGES.rst:
```diff
@@ -4340,6 +4340,10 @@ astropy.coordinates
 - The ``__repr__`` of coordinate objects now shows scalar coordinates in the
   same format as vector coordinates. [#3350, 3448]
+  
+- The ``lookup_name`` method of the ``TransformGraph`` class now has an extra
```
Could you move this item to the version 3.0 section of coordinates?

MSeifert04 commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,35 @@ def get_transform(self, fromsys, tosys):
         ----------
         name : str
             The alias to look up.
+        suppress_error : bool (..versionadded:: 3.0)
```
Ah, the `.. versionadded:: 3.0` is a paragraph directive. So you have to put it after the text on a newline:

```
suppress_error : bool (..versionadded:: 3.0)
    If True (the default) then it returns None if the name isn't 
    found otherwise it will raise an Exception.
   
    .. versionadded:: 3.0
```

I would really recommend you try to build the documentation locally and verify that it, indeed, looks okay :)

MSeifert04 commented:
Amending is not necessary, that's why I mentioned squashing. If you're interested I could give you the necessary commands or provide some more verbose step-by-step tutorials.

ayush1999 commented:
@MSeifert04 . I'd love to get guidance on how squashing works. Can I please contact you via gitter? Thank you for your support! :)

MSeifert04 commented on astropy/coordinates/transformations.py:
```diff
@@ -347,23 +347,37 @@ def get_transform(self, fromsys, tosys):
+            If True (the default) then it returns None if the name isn't
+            found otherwise it will raise an Exception.
+
+            ..versionadded:: 3.0
```
Needs a whitespace between `..` and `versionadded`

MSeifert04 commented:
I'm a bit busy right now, but I probably can join the channel later. But just a more general question: Do you use git from the command line or via some GUI?

ayush1999 commented:
@MSeifert04 I've made the final changes. Also, I prefer using the terminal, but I find using GitHub directly to be easier sometimes.

ayush1999 commented:
Can someone please tell me why are two of the Comprehensive tests in travis failing?

ayush1999 commented:
@MSeifert04 Can you please review this? :)

astrofrog commented:
I'll leave this closed - if someone wants to implement this, I'd recommend opening a new PR with just the required commits from here plus the additional changes requested in the the above reviews.
